--- Library/Homebrew/formula_installer.rb      2018-12-07 09:00:49.000000000 -0600
+++ Library/Homebrew/formula_installer.rb 2018-12-26 22:02:04.000000000 -0600
@@ -313,6 +313,8 @@

     build_bottle_preinstall if build_bottle?

+    return if ENV['HOMEBREW_DOWNLOAD_ONLY'] != ''
+
     unless @poured_bottle
       build
       clean
@@ -618,6 +620,8 @@
   end

   def finish
+    return if ENV['HOMEBREW_DOWNLOAD_ONLY'] != ''
+
     return if only_deps?

     ohai "Finishing up" if verbose?
